### PR TITLE
AB#101536: Client Auth

### PR DESCRIPTION
**Pull request by ShasonGurung**
## Overview

- Login and Registration screens including post and response handling.
- Update main menu welcome to be users name on successful login.
- Automatically log in the user if requirement met or else display Log in screen with an option to register

**Registration**
image image image

**Home Screen**
image

## Azure Boards

- AB#101546
- AB#101547
- AB#101549
- AB#101550
- AB#101556
- AB#101557

## Notes

- Since there's no mechanism on RN where cookies are automatically stored, an interceptor has been setup. This intercepts an API request and response before reaching its final destination. Here, it does the two jobs. Firstly, intercepts API response and extract relevant cookies if available and then stores them in Async Storage, which is similar to Local Storage. Secondly, it intercepts API request and attach cookies if available in the Async Storage into the request header before sending request.

### CODE REVIEW

File: app/react-native-app/src/components/PasswordField.jsx [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+      /[A-Z]/,
+      "Passwords must have at least one uppercase character ('A' - 'Z')."
+    )
+    .matches(
+      /[a-z]/,
+      "Passwords must have at least one lower character ('a' - 'z')."
+    )
+    .matches(
+      /[^A-Za-z0-9]/,
+      "Passwords must have at least one non alphanumeric character."
```
Comment (dcl10) [inline app/react-native-app/src/components/PasswordField.jsx:21]:
```suggestion
      "Passwords must have at least one non-alphanumeric character."
```
⛏️

File: app/react-native-app/src/components/PasswordField.jsx [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+      minLength,
+    })
+    .matches(/\d/, "Passwords must have at least one digit ('0' - '9').")
+    .matches(
+      /[A-Z]/,
+      "Passwords must have at least one uppercase character ('A' - 'Z')."
+    )
+    .matches(
+      /[a-z]/,
+      "Passwords must have at least one lower character ('a' - 'z')."
```
Comment (dcl10) [inline app/react-native-app/src/components/PasswordField.jsx:17]:
```suggestion
      "Passwords must have at least one lowercase character ('a' - 'z')."
```
⛏️

File: app/react-native-app/src/constants/account-registration.js [hunk 3]
```diff
@@ -19,12 +19,16 @@ export const accountRegistration = {
       label: "Email",
       defaultValue: "",
     },
+    isVeterinarian: {
+      label: "Are you registering as Veterinarian ?",
```
Comment (dcl10) [inline app/react-native-app/src/constants/account-registration.js:23]:
```suggestion
      label: "Are you registering as Veterinarian?",
```
⛏️

File: app/react-native-app/src/contexts/Config.jsx [hunk 4]
```diff
@@ -1,13 +1,24 @@
...
+    const value = await AsyncStorage.getItem(".EquineCastration.Config");
+    if (value !== null) {
+      const decodedCookie = decodeURIComponent(value.split("=")[1]);
+      return JSON.parse(decodedCookie);
+    }
+  } catch (e) {
+    console.log(
+      "Error reading cookie '.EquineCastration.Config' from Async Storage",
+      e
+    );
```
Comment (dcl10) [inline app/react-native-app/src/contexts/Config.jsx:19]: If this is for debugging purposes, just replace this with `return null` and handle the null case in the calling code.

File: app/react-native-app/src/contexts/User.jsx [hunk 5]
```diff
@@ -1,35 +1,67 @@
...
-};
+// Clear the entire AsyncStorage ()
+// const clearAsyncStorage = async () => {
+//   try {
+//     await AsyncStorage.clear();
+//     console.log("AsyncStorage cleared!");
+//   } catch (e) {
+//     console.log("Error clearing AsyncStorage: ", e);
+//   }
+// };
```
Comment (dcl10) [inline app/react-native-app/src/contexts/User.jsx:17]: Delete if you don't need it anymore.

File: app/react-native-app/src/screens/Account/Registration.jsx [hunk 6]
```diff
@@ -51,36 +59,294 @@ const InitialValues = (keysArr, store) => {
...
-    <Formik
-      initialValues={initialValues}
-      validationSchema={validationSchema()}
-      onSubmit={async (values) => {
-        console.log(values);
-        // TODO: Handle submission
-      }}
+    <AccountLayout
+      primaryHeading="GDPR Consent Form"
+      secondaryHeading="Please carefully read the GDPR consent"
```
Comment (dcl10) [inline app/react-native-app/src/screens/Account/Registration.jsx:270]:
```suggestion
      secondaryHeading="Please carefully read the GDPR consent form"
```

**Review by dcl10 — CHANGES_REQUESTED** [review]

File: app/react-native-app/src/screens/UserHome.jsx [hunk 7]
```diff
@@ -70,7 +70,7 @@ export const UserHome = ({ navigation }) => {
               color: colors.primary[700],
             }}
           >
-            {user?.fullName.toUpperCase()}
+            {user?.fullName?.toUpperCase()}
```
Comment (dcl10) [inline app/react-native-app/src/screens/UserHome.jsx:73]: Why would we need to check if the user be null or undefined on the user screen? Doesn't a user have to logged in to view this screen?

**Review by dcl10 — APPROVED** [review]

**Review by AndyRae — APPROVED** [review]

**Pull request merged**